src/main/java/com/google/cloud/tools/appengine/cloudsdk/GcloudRunner.java

```diff
@@ -0,0 +1,150 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.appengine.cloudsdk;
+
+import com.google.cloud.tools.appengine.cloudsdk.internal.args.GcloudArgs;
+import com.google.cloud.tools.appengine.cloudsdk.internal.process.ProcessBuilderFactory;
+import com.google.cloud.tools.appengine.cloudsdk.process.ProcessHandler;
+import com.google.cloud.tools.appengine.cloudsdk.process.ProcessHandlerException;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Joiner;
+import com.google.common.collect.Maps;
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+
+class GcloudRunner {
+
+  private static final Logger logger = Logger.getLogger(GcloudRunner.class.getName());
+
+  private final CloudSdk sdk;
+  private final String metricsEnvironment;
+  private final String metricsEnvironmentVersion;
+  private final File credentialFile;
+  private final String outputFormat;
+  private final String showStructuredLogs;
+  private final ProcessBuilderFactory processBuilderFactory;
+  private final ProcessHandler processHandler;
+
+  GcloudRunner(
+      CloudSdk sdk,
+      String metricsEnvironment,
+      String metricsEnvironmentVersion,
+      File credentialFile,
+      String outputFormat,
+      String showStructuredLogs,
+      ProcessBuilderFactory processBuilderFactory,
+      ProcessHandler processHandler) {
+    this.sdk = sdk;
+    this.metricsEnvironment = metricsEnvironment;
+    this.metricsEnvironmentVersion = metricsEnvironmentVersion;
+    this.credentialFile = credentialFile;
+    this.outputFormat = outputFormat;
+    this.showStructuredLogs = showStructuredLogs;
+    this.processBuilderFactory = processBuilderFactory;
+    this.processHandler = processHandler;
+  }
+
+  void run(List<String> arguments, File workingDirectory)
+      throws ProcessHandlerException, CloudSdkNotFoundException, CloudSdkOutOfDateException,
+          CloudSdkVersionFileException, InvalidJavaSdkException, IOException {
+
+    sdk.validateCloudSdk();
+
+    List<String> command = new ArrayList<>();
+    command.add(sdk.getGCloudPath().toString());
+
+    command.addAll(arguments);
+    command.addAll(GcloudArgs.get("format", outputFormat));
+
+    if (credentialFile != null) {
+      command.addAll(GcloudArgs.get("credential-file-override", credentialFile));
+    }
+
+    logger.info("submitting command: " + Joiner.on(" ").join(command));
+
+    ProcessBuilder processBuilder = processBuilderFactory.newProcessBuilder();
+    processBuilder.command(command);
+    if (workingDirectory != null) {
```

Comment:
`null` check shouldn't be necessary since `directory` can take `null`

Reply:
updated